parcels/spatialhash.py

```diff
@@ -30,171 +30,234 @@ def __init__(
+            # Boundaries of the hash grid are the unit cube
+            self._xmin = -1.0
+            self._ymin = -1.0
+            self._zmin = -1.0
+            self._xmax = 1.0
+            self._ymax = 1.0
+            self._zmax = 1.0  # Compute the cell centers of the source grid (for now, assuming Xgrid)
+            lon = np.deg2rad(self._source_grid.lon)
+            lat = np.deg2rad(self._source_grid.lat)
+            x, y, z = _latlon_rad_to_xyz(lat, lon)
+            _xbound = np.stack(
+                (
+                    x[:-1, :-1],
+                    x[:-1, 1:],
+                    x[1:, 1:],
+                    x[1:, :-1],
+                ),
+                axis=-1,
             )
-            coords = np.stack(
+            _ybound = np.stack(
                 (
-                    lat_bounds[:, :, 0].flatten(),
-                    lon_bounds[:, :, 0].flatten(),
+                    y[:-1, :-1],
+                    y[:-1, 1:],
+                    y[1:, 1:],
+                    y[1:, :-1],
                 ),
                 axis=-1,
             )
-            yi1, xi1 = self._hash_index2d(coords)
-            coords = np.stack(
+            _zbound = np.stack(
                 (
-                    lat_bounds[:, :, 1].flatten(),
-                    lon_bounds[:, :, 1].flatten(),
+                    z[:-1, :-1],
+                    z[:-1, 1:],
+                    z[1:, 1:],
+                    z[1:, :-1],
                 ),
                 axis=-1,
             )
-            yi2, xi2 = self._hash_index2d(coords)
-            nface = (self._source_grid.xdim) * (self._source_grid.ydim)
-            for eid in range(nface):
-                for j in range(yi1[eid], yi2[eid] + 1):
-                    if xi1[eid] <= xi2[eid]:
-                        # Normal case, no wrap
-                        for i in range(xi1[eid], xi2[eid] + 1):
-                            index_to_face[(i % self._nx) + self._nx * j].append(eid)
-                    else:
-                        # Wrap-around case
-                        for i in range(xi1[eid], self._nx):
-                            index_to_face[(i % self._nx) + self._nx * j].append(eid)
-                        for i in range(0, xi2[eid] + 1):
-                            index_to_face[(i % self._nx) + self._nx * j].append(eid)
-            return index_to_face
+            # Compute centroid locations of each cells
+            self._xc = np.mean(_xbound, axis=-1)
+            self._yc = np.mean(_ybound, axis=-1)
+            self._zc = np.mean(_zbound, axis=-1)
+
+        else:
+            # Boundaries of the hash grid are the bounding box of the source grid
+            self._xmin = self._source_grid.lon.min()
+            self._xmax = self._source_grid.lon.max()
+            self._ymin = self._source_grid.lat.min()
+            self._ymax = self._source_grid.lat.max()
+            self._zmin = 0.0
+            self._zmax = 0.0
```

Comment:
Do we even need `z`, if we are 2D and it appears to be zero everywhere? Or do I miss something? 
Because if this is to generalise later to 3dimensional search; I don't think we need that. I don't expect us to do any 3D searches, given that depth is always perpendicular to lon/lat in the grid (and thus can be searched in 1D)

Reply:
OK clear. Perhaps good to add this a comment that this is needed for `mesh="flat"`?
```suggestion
            # setting min and max below is needed for mesh="flat"
            self._zmin = 0.0
            self._zmax = 0.0
```